File: archive-triage-runner/python/src/archive_triage/gedcom_scanner.py
"""Walk a directory and yield ``(path, content)`` pairs for every ``*.ged`` file."""

from __future__ import annotations

from collections.abc import Iterator
from dataclasses import dataclass
from pathlib import Path

_GEDCOM_SUFFIX = ".ged"


class ScannerError(ValueError):
    """Raised when the input directory is unusable."""


@dataclass(frozen=True)
class GedcomFile:
    path: Path
    content: str


class GedcomScanner:
    def __init__(self, root: Path) -> None:
        self._root = root

    def scan(self) -> Iterator[GedcomFile]:
        if not self._root.is_dir():
            raise ScannerError(
                f"Input path is not a directory: {self._root} "
                "(pass a folder that contains *.ged files)."
            )
        for path in sorted(self._root.rglob(f"*{_GEDCOM_SUFFIX}")):
            if path.is_file():
                yield GedcomFile(path=path, content=path.read_text(encoding="utf-8"))

File: archive-triage-runner/dotnet/src/ArchiveTriage/GedcomScanner.cs
namespace ArchiveTriage;

public sealed record GedcomFile(string Path, string Content);

public sealed class ScannerException : Exception
{
    public ScannerException(string message) : base(message) { }
}

public sealed class GedcomScanner
{
    private const string GedcomSuffix = ".ged";

    private readonly string root;

    public GedcomScanner(string root)
    {
        this.root = root;
    }

    public IEnumerable<GedcomFile> Scan()
    {
        if (!Directory.Exists(this.root))
        {
            throw new ScannerException(
                $"Input path is not a directory: {this.root} "
                + "(pass a folder that contains *.ged files).");
        }

        var files = Directory.EnumerateFiles(this.root, $"*{GedcomSuffix}", SearchOption.AllDirectories)
            .OrderBy(path => path, StringComparer.Ordinal);

        foreach (var path in files)
        {
            yield return new GedcomFile(path, File.ReadAllText(path));
        }
    }
}
